### PR TITLE
Add gcom api processor.

### PR DESCRIPTION
**Pull request by kovrus**
See [Design Doc: Use OTel Collector as a replacement for OTLP Gateway in Grafana Cloud OTLP Endpoint
](https://docs.google.com/document/d/1HsJr5eVH4WOdSSGIeaYRRUGAqx4Bzx-CSna1mspW4a4/edit#heading=h.5sybau7waq2q) for more information on why this component is needed.

Note: I've copied only cache and client modules from `https://github.com/grafana/backend-enterprise`. I did not use it as dep because it causes some des conflicts and it has to be fetched from private repo (which might be ok actually).   

Follow up:
- add `otlp-gateway` distribution with this and the routing processor component. ( https://github.com/grafana/opentelemetry-collector-components/pull/46/commits/fa135124bff990305a487e7692319ab3704f0ad1)

### CODE REVIEW

**Comment by kovrus** [comment]
`internal.{cache, client, common}` should not be reviewed, it is one to one copy from  `https://github.com/grafana/backend-enterprise`

File: components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/opentelemetry-collector-components/processor/gcomapiprocessor/internal/gcom/client"
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go:11]: we might need a linter to enforce three import groups -- this one here would belong to a third group, of local imports

File: components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+func (a *AuthCache) Clear() {
+	a.itemsMtx.Lock()
+	defer a.itemsMtx.Unlock()
+
+	a.items = make(map[string]cacheItem)
+}
+
+// CheckKey checks if the provided key is valid
+func (a *AuthCache) CheckKey(ctx context.Context, key string) (*client.APIKey, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "Gateway.CheckKey")
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go:67]: I know which code I'll change from opentracing to otel first :-)

File: components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/go-kit/log/level"
+	"github.com/grafana/opentelemetry-collector-components/processor/gcomapiprocessor/internal/gcom/client"
+	"github.com/opentracing/opentracing-go"
+)
+
+var errBadKey = errors.New("API Key Not Valid")
+
+type AuthCache struct {
+	client   client.Client
+	logger   log.Logger
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go:19]: The collerctor uses zap instead -- does it make sense to switch to it at a follow-up PR?

File: components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+		if err != nil {
+			return err
+		}
+		*it = append(*it, instanceType)
+	}
+	return nil
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *InstanceCacheConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go:50]: Are those flags actually used?

File: components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+func NewInstanceCache(
+	cfg InstanceCacheConfig,
+	logger log.Logger,
+	instanceTypes []client.InstanceType,
+	gclient client.Client,
+) (InstanceCache, error) {
+
+	if len(instanceTypes) == 0 {
+		// TODO (@chroberts): replace this with an error after we roll out
+		// https://github.com/grafana/backend-enterprise/pull/3749
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go:107]: This is merged -- should this code be changed already?

File: components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+		level.Debug(i.logger).Log("msg", "adding/updating instance to cache", "instanceID", instance.ID)
+		i.instances[instanceType][instance.ID] = instance
+		i.idByOrgIDAndName[instanceType][strconv.Itoa(instance.OrgID)+instance.Name] = instance.ID
+	}
+
+	return nil
+}
+
+func (i *instanceCache) run() {
+	completeRefreshTicker := time.NewTicker(i.cfg.CompleteCacheRefreshDuration)
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go:228]: Ideally, this would receive a context, which would be cancelled when the component is shutting down. In that case, the for range channel would be replaced with selects, so that we can cancel the cache refresh operations when we are shutting down.

File: components/processor/gcomapiprocessor/internal/gcom/client/api_key.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+		strings.NewReader(payload.Encode()),
+	)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("err=%w, msg=%v", ErrAccess, err)
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/client/api_key.go:65]: Are we leaking internal details with this message? Or do we trust that the internal gateway is returning only information that is safe for the general public?

File: components/processor/gcomapiprocessor/internal/gcom/client/client.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/opentelemetry-collector-components/processor/gcomapiprocessor/internal/gcom/common"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	grafanaComReqs = promauto.NewHistogramVec(
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/client/client.go:20]: TODO: change this to otel, not sure if those would join the ones that are generated by the collector itself. In the future, the collector might not export automatically to Prometheus like it does today.

File: components/processor/gcomapiprocessor/internal/gcom/client/client.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	Timeout  time.Duration
+	Client   *http.Client
+
+	// Mock allows for a mock client to be substituted
+	// and returned by the New function
+	Mock Client
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/client/client.go:44]: Should those flags be removed?

File: components/processor/gcomapiprocessor/internal/gcom/client/client.go (hunk 8 above)
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/client/client.go:20]: Do we want to use the telemetry settings that the collector provides? While the registries might end up merged, this is likely to change in the future.

File: components/processor/gcomapiprocessor/internal/gcom/client/client.go (hunk 9 above)
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/client/client.go:44]: Should we remove this?

File: components/processor/gcomapiprocessor/internal/gcom/client/client.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+type client struct {
+	key      string
+	endpoint *url.URL
+	client   *http.Client
+	logger   log.Logger
+}
+
+// New creates a grafana.com Client
+func New(cfg Config, name string, logger log.Logger) (Client, error) {
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/internal/gcom/client/client.go:78]: I just realized that those are the two packages you said you just copied from the enterprise package. If we are going to maintain them, should we accomodate them to the collector standards? Or do we want to threat those two packages as libraries? In that case, it would probably make sense to split them into different go modules perhaps.

File: components/processor/gcomapiprocessor/testdata/config.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+gcomapi:
+  client:
+    endpoint: "http://localhost:3000"
+    key: "test_key"
+    timeout: 1s
+  cache:
+    complete_cache_refresh_duration: 1s
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/testdata/config.yaml:7]: complete_refresh_duration -- "cache" is already the parent's node name

File: components/processor/gcomapiprocessor/testdata/config.yaml [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    incremental_cache_refresh_duration: 1s
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/testdata/config.yaml:8]: incremental_refresh_duration

File: components/processor/gcomapiprocessor/README.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+      processors: [ gcomapi, routing ]
+      exporters: [ logging/1 ]
+    metrics:
+      receivers: [ otlp ]
+      processors: [ gcomapi, routing ]
+      exporters: [ logging/1 ]
+```
+
+[alpha]: https://github.com/open-telemetry/opentelemetry-collector#alpha
+[design doc]: https://docs.google.com/document/d/1HsJr5eVH4WOdSSGIeaYRRUGAqx4Bzx-CSna1mspW4a4/edit#heading=h.89ldx0hih690
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/README.md:67]: This is Grafana-internal. Is this a problem? Not sure myself.

File: components/processor/gcomapiprocessor/config.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type Config struct {
+	serviceName string       `mapstructure:"service_name"`
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/config.go:11]: Shouldn't this be in upper case?

File: components/processor/gcomapiprocessor/config.go [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+	Client      clientConfig `mapstructure:"client"`
+	Cache       cacheConfig  `mapstructure:"cache"`
+	IsTest      bool         `mapstructure:"test"`
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/config.go:14]: Dry-run? Mock? Dev? Perhaps we don't even need this if we provide a container image with a gcom API mock?

File: components/processor/gcomapiprocessor/go.mod [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+	github.com/stretchr/testify v1.8.1
+	go.opentelemetry.io/collector v0.70.0
+	go.opentelemetry.io/collector/component v0.70.0
+	go.opentelemetry.io/collector/confmap v0.70.0
+	go.opentelemetry.io/collector/consumer v0.70.0
+	go.opentelemetry.io/collector/pdata v1.0.0-rc4
+	go.uber.org/zap v1.24.0
+)
+
+require (
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/go.mod:15]: Why do we have three require groups?

File: components/processor/gcomapiprocessor/config.go (hunk 15 above)
Comment (jpkrohling) [reply components/processor/gcomapiprocessor/config.go:14]: Or perhaps we could have a special handling for an endpoint that starts with "mock://" ?

File: components/processor/gcomapiprocessor/processor.go [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+			Key:      cfg.Client.Endpoint,
+			Timeout:  cfg.Client.Timeout,
+		},
+		cfg.serviceName,
+		logger,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if cfg.IsTest {
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/processor.go:51]: Here we could check for the "mock://" ?

File: components/processor/gcomapiprocessor/processor.go [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+	ctx context.Context,
+	extractURL func(i client.Instance) string,
+) (context.Context, error) {
+
+	// Extract X-Scope-OrgID
+	info := collectorclient.FromContext(ctx)
+	v := info.Metadata.Get(orgID)
+
+	if v == nil || len(v) == 0 {
+		return nil, fmt.Errorf("missing '%s' header", orgID)
```
Comment (jpkrohling) [inline components/processor/gcomapiprocessor/processor.go:86]: You could have used %q here: "missing %q header"

File: go.sum [hunk 19]
```diff
@@ -2,8 +2,6 @@ github.com/caarlos0/log v0.1.2 h1:P/pppG5Pv2wfapmioBCQaydFUteMbOy9nSttUHbdowY=
 github.com/caarlos0/log v0.1.2/go.mod h1:lYxaBNu0NYLm5tdxBysIb2LNhNUUFqNAzSHNu737Loo=
 github.com/caarlos0/testfs v0.4.4 h1:3PHvzHi5Lt+g332CiShwS8ogTgS3HjrmzZxCm6JCDr8=
 github.com/caarlos0/testfs v0.4.4/go.mod h1:bRN55zgG4XCUVVHZCeU+/Tz1Q6AxEJOEJTliBy+1DMk=
-github.com/charmbracelet/lipgloss v0.5.1-0.20220615005615-2e17a8a06096 h1:ai19sA3Zyg3DARevWCbdLOWt+MfWiE3e8voBqzFOgP8=
-github.com/charmbracelet/lipgloss v0.5.1-0.20220615005615-2e17a8a06096/go.mod h1:D7uPgcyfB9T1Ug2mfJOnES17o47nz5oqIzSSVrpcviU=
```
Comment (jpkrohling) [inline go.sum:6]: This does not seem related?

**Review by jpkrohling — APPROVED** [review]
I'm OK with merging this as is, as I believe you tested and said this worked already. Everything I listed here can be done later on follow-up PRs.

File: components/processor/gcomapiprocessor/internal/gcom/client/client.go (hunk 10 above)
Comment (kovrus) [reply components/processor/gcomapiprocessor/internal/gcom/client/client.go:78]: I'll follow up with a PR against backend-enterprise to convert those packages to modules. If it wouldn't work for some reason, let's improve the code and strip all unnecessary parts.

File: components/processor/gcomapiprocessor/README.md (hunk 13 above)
Comment (kovrus) [reply components/processor/gcomapiprocessor/README.md:67]: The component is Grafana internal as well, so linking this design doc should be fine I guess.

File: components/processor/gcomapiprocessor/config.go (hunk 15 above)
Comment (kovrus) [reply components/processor/gcomapiprocessor/config.go:14]: > perhaps we don't even need this if we provide a container image with a gcom API mock?

> Or perhaps we could have a special handling for an endpoint that starts with "mock://" ?

I like both options, but would probably lean towards `mock://` since the test setup will be simpler.

**Comment by kovrus** [comment]
I've addressed all the comments except comments on the `internal` package, I'd like to try to create client and cache modules in backend enterprise first, if it doesn't work, I'll address those comments too in a separate PR. For now, I'll proceed with merging this PR.
